tools/crm_resource_ban.c

```diff
@@ -235,3 +238,135 @@ cli_resource_clear(const char *rsc_id, const char *host, GListPtr allnodes, cib_
     free_xml(fragment);
     return rc;
 }
+
+static char *
+build_clear_xpath_string(xmlNode *constraint_node, const char *rsc, const char *node, bool scope_master)
+{
+    int offset = 0;
+    char *xpath_string = NULL;
+    char *first_half = NULL;
+    char *rsc_role_substr = NULL;
+    char *date_substr = NULL;
+
+    if (crm_starts_with(ID(constraint_node), "cli-ban-")) {
+        date_substr = crm_strdup_printf("//date_expression[@id='%s-lifetime']",
+                                        ID(constraint_node));
+
+    } else if (crm_starts_with(ID(constraint_node), "cli-prefer-")) {
+        date_substr = crm_strdup_printf("//date_expression[@id='cli-prefer-lifetime-end-%s']",
+                                        crm_element_value(constraint_node, "rsc"));
+    } else {
+        return NULL;
+    }
+
+    first_half = calloc(1, XPATH_MAX);
+    offset += snprintf(first_half + offset, XPATH_MAX - offset, "//rsc_location");
+
+    if (node != NULL || rsc != NULL || scope_master == TRUE) {
+        offset += snprintf(first_half + offset, XPATH_MAX - offset, "[");
+
+        if (node != NULL) {
+            if (rsc != NULL || scope_master == TRUE) {
+                offset += snprintf(first_half + offset, XPATH_MAX - offset, "@node='%s' and ", node);
+            } else {
+                offset += snprintf(first_half + offset, XPATH_MAX - offset, "@node='%s'", node);
+            }
+        }
+
+        if (rsc != NULL && scope_master == TRUE) {
+            rsc_role_substr = crm_strdup_printf("@rsc='%s' and @role='%s'", rsc, RSC_ROLE_MASTER_S);
+            offset += snprintf(first_half + offset, XPATH_MAX - offset, "@rsc='%s' and @role='%s']", rsc, RSC_ROLE_MASTER_S);
+        } else if (rsc != NULL) {
+            rsc_role_substr = crm_strdup_printf("@rsc='%s'", rsc);
+            offset += snprintf(first_half + offset, XPATH_MAX - offset, "@rsc='%s']", rsc);
+        } else if (scope_master == TRUE) {
+            rsc_role_substr = crm_strdup_printf("@role='%s'", RSC_ROLE_MASTER_S);
+            offset += snprintf(first_half + offset, XPATH_MAX - offset, "@role='%s']", RSC_ROLE_MASTER_S);
+        } else {
+            offset += snprintf(first_half + offset, XPATH_MAX - offset, "]");
+        }
+    }
+
+    if (node != NULL) {
+        if (rsc_role_substr != NULL) {
+            xpath_string = crm_strdup_printf("%s|//rsc_location[%s]/rule[expression[@attribute='#uname' and @value='%s']]%s",
+                                             first_half, rsc_role_substr, node, date_substr);
+        } else {
+            xpath_string = crm_strdup_printf("%s|//rsc_location/rule[expression[@attribute='#uname' and @value='%s']]%s",
+                                             first_half, node, date_substr);
+        }
+    } else {
+        xpath_string = crm_strdup_printf("%s%s", first_half, date_substr);
+    }
+
+    free(first_half);
+    free(date_substr);
+    free(rsc_role_substr);
+
+    return xpath_string;
+}
```

Comment:
lol, I hope you don't feel too dirty after writing this function :) xpaths are powerful but a pain to work with

Reply:
Hah, yeah, it was not a lot of fun to write.  But yeah - xpath is really useful.  I'd rather write that one query than write all the code necessary to do the searching by hand.